swift/ql/test/query-tests/Security/CWE-328/testCrypto.swift

```diff
@@ -21,48 +52,119 @@ enum Insecure {
+    hash.update(data: passwd)  // BAD [NOT DETECTED] not a computationally hard hash
+    hash.update(data: cert)  // GOOD
+    hash.update(data: account_no)   // GOOD
+    hash.update(data: credit_card_no)   // GOOD
+}
+
+func testMD5UpdateWithUnsafeRawBufferPointer(passwd : UnsafeRawBufferPointer, cert: UnsafeRawBufferPointer, encrypted_passwd : UnsafeRawBufferPointer, account_no : UnsafeRawBufferPointer, credit_card_no : UnsafeRawBufferPointer) {
     var hash = Crypto.Insecure.MD5()
     hash.update(bufferPointer: passwd)  // BAD
+    hash.update(bufferPointer: cert)  // BAD [NOT DETECTED]
     hash.update(bufferPointer: encrypted_passwd)  // GOOD  (not sensitive)
     hash.update(bufferPointer: account_no)   // BAD [NOT DETECTED]
     hash.update(bufferPointer: credit_card_no)   // BAD
 }
 
-func test6(passwd : UnsafeRawBufferPointer, encrypted_passwd : UnsafeRawBufferPointer, account_no : UnsafeRawBufferPointer, credit_card_no : UnsafeRawBufferPointer) {
+func testSHA1UpdateWithUnsafeRawBufferPointer(passwd : UnsafeRawBufferPointer, cert: UnsafeRawBufferPointer, encrypted_passwd : UnsafeRawBufferPointer, account_no : UnsafeRawBufferPointer, credit_card_no : UnsafeRawBufferPointer) {
     var hash = Crypto.Insecure.SHA1()
     hash.update(bufferPointer: passwd)  // BAD
+    hash.update(bufferPointer: cert)  // BAD [NOT DETECTED]
     hash.update(bufferPointer: encrypted_passwd)  // GOOD  (not sensitive)
     hash.update(bufferPointer: account_no)   // BAD [NOT DETECTED]
     hash.update(bufferPointer: credit_card_no)   // BAD
 }
+
+func testSHA256UpdateWithUnsafeRawBufferPointer(passwd : UnsafeRawBufferPointer, cert: UnsafeRawBufferPointer, encrypted_passwd : UnsafeRawBufferPointer, account_no : UnsafeRawBufferPointer, credit_card_no : UnsafeRawBufferPointer) {
+    var hash = Crypto.SHA256()
+    hash.update(bufferPointer: passwd)  // BAD [NOT DETECTED] not a computationally hard hash
+    hash.update(bufferPointer: cert)  // GOOD
+    hash.update(bufferPointer: account_no)   // GOOD
+    hash.update(bufferPointer: credit_card_no)   // GOOD
+}
+
+func testSHA384UpdateWithUnsafeRawBufferPointer(passwd : UnsafeRawBufferPointer, cert: UnsafeRawBufferPointer, encrypted_passwd : UnsafeRawBufferPointer, account_no : UnsafeRawBufferPointer, credit_card_no : UnsafeRawBufferPointer) {
+    var hash = Crypto.SHA384()
+    hash.update(bufferPointer: passwd)  // BAD [NOT DETECTED] not a computationally hard hash
+    hash.update(bufferPointer: cert)  // GOOD
+    hash.update(bufferPointer: account_no)   // GOOD
+    hash.update(bufferPointer: credit_card_no)   // GOOD
+}
+
+func testSHA512UpdateWithUnsafeRawBufferPointer(passwd : UnsafeRawBufferPointer, cert: UnsafeRawBufferPointer, encrypted_passwd : UnsafeRawBufferPointer, account_no : UnsafeRawBufferPointer, credit_card_no : UnsafeRawBufferPointer) {
+    var hash = Crypto.SHA512()
+    hash.update(bufferPointer: passwd)  // BAD [NOT DETECTED] not a computationally hard hash
```

Comment:
What *is* good hashing for a password?

Reply:
any hash that is somewhat hard to compute, in a way that is ok for normal usage but precludes computing billions of hashes in a reasonable time. There are more details on this OWASP cheat sheet page (that is linked in the qhelp file :slightly_smiling_face:):

https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html

Reply:
that's something that I did not know and actually surprised me, and that I learned from the python/ruby/go version of the query

Reply:
Thanks, I see what's going on now.

I guess it isn't an issue right now, because we only flag MD5 and SHA1 - but if at some point we flag password hashing using stronger algorithms, I think we'll want to have a test and an example in the qhelp showing exactly how to do it correctly.

To be clear, this is not a change request for this PR.